File: src/webrtc.cpp
/*
 *  Copyright © 2017-2020 Wellington Wallace
 *
 *  This file is part of EasyEffects.
 *
 *  EasyEffects is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  EasyEffects is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with EasyEffects.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "webrtc.hpp"

// namespace {

// void on_n_input_samples_changed(GObject* gobject, GParamSpec* pspec, Webrtc* w) {
//   int v = 0;
//   int blocksize = 0;

//   g_object_get(w->adapter, "n-input-samples", &v, nullptr);
//   g_object_get(w->adapter, "blocksize", &blocksize, nullptr);

//   util::debug(w->log_tag + "webrtc: new input block size " + std::to_string(v) + " frames");
//   util::debug(w->log_tag + "webrtc: we will try to read in chunks of " + std::to_string(blocksize) + " frames");

//   g_object_set(w->adapter_out, "blocksize", v, nullptr);
// }

// }  // namespace

Webrtc::Webrtc(const std::string& tag,
               const std::string& schema,
               const std::string& schema_path,
               const int& sampling_rate,
               PipeManager* pipe_manager)
    : PluginBase(tag, "webrtc", schema, schema_path, pipe_manager), rate(sampling_rate) {
  // webrtc = gst_element_factory_make("webrtcdsp", nullptr);

  // if (is_installed(webrtc)) {
  //   build_probe_bin();
  //   build_dsp_bin();

  //   bind_to_gsettings();

  //   // useless write just to force callback call

  //   auto enable = g_settings_get_boolean(settings, "state");

  //   g_settings_set_boolean(settings, "state", enable);
  // }
}

Webrtc::~Webrtc() {
  util::debug(log_tag + name + " destroyed");
}

void Webrtc::build_probe_bin() {
  // probe_bin = gst_bin_new("probe_bin");

  // probe_src = gst_element_factory_make("pipewiresrc", nullptr);
  // auto* queue = gst_element_factory_make("queue", nullptr);
  // auto* audioconvert = gst_element_factory_make("audioconvert", nullptr);
  // auto* audioresample = gst_element_factory_make("audioresample", nullptr);
  // auto* capsfilter = gst_element_factory_make("capsfilter", nullptr);
  // auto* probe = gst_element_factory_make("webrtcechoprobe", nullptr);
  // auto* sink = gst_element_factory_make("fakesink", nullptr);

  // auto* props = gst_structure_from_string("props,application.name=EasyEffectsWebrtcProbe", nullptr);

  // const auto* caps_str = "audio/x-raw,format=S16LE,channels=2,rate=48000";
  // auto* caps = gst_caps_from_string(caps_str);

  // g_object_set(probe_src, "always-copy", 1, nullptr);
  // g_object_set(probe_src, "do-timestamp", 1, nullptr);
  // g_object_set(probe_src, "stream-properties", props, nullptr);

  // g_object_set(capsfilter, "caps", caps, nullptr);

  // g_object_set(queue, "silent", 1, nullptr);

  // gst_structure_free(props);
  // gst_caps_unref(caps);

  // gst_bin_add_many(GST_BIN(probe_bin), probe_src, queue, audioconvert, audioresample, capsfilter, probe, sink,
  // nullptr);

  // gst_element_link_many(probe_src, queue, audioconvert, audioresample, capsfilter, probe, sink, nullptr);
}

void Webrtc::build_dsp_bin() {
  // auto* in_level = gst_element_factory_make("level", "webrtc_input_level");
  // auto* audioconvert_in = gst_element_factory_make("audioconvert", nullptr);
  // auto* audioresample_in = gst_element_factory_make("audioresample", nullptr);
  // auto* caps_in = gst_element_factory_make("capsfilter", nullptr);
  // auto* audioconvert_out = gst_element_factory_make("audioconvert", nullptr);
  // auto* audioresample_out = gst_element_factory_make("audioresample", nullptr);
  // auto* caps_out = gst_element_factory_make("capsfilter", nullptr);
  // auto* out_level = gst_element_factory_make("level", "webrtc_output_level");

  // adapter = gst_element_factory_make("peadapter", nullptr);
  // adapter_out = gst_element_factory_make("peadapter", nullptr);

  // auto* capsin = gst_caps_from_string("audio/x-raw,channels=2,format=S16LE,rate=48000");
  // auto* capsout = gst_caps_from_string(("audio/x-raw,channels=2,format=F32LE,rate=" + std::to_string(rate)).c_str());

  // g_object_set(caps_in, "caps", capsin, nullptr);
  // g_object_set(caps_out, "caps", capsout, nullptr);

  // gst_caps_unref(capsin);
  // gst_caps_unref(capsout);

  // g_object_set(adapter, "blocksize", 480, nullptr);

  // gst_bin_add(GST_BIN(bin), probe_bin);

  // gst_bin_add_many(GST_BIN(bin), in_level, adapter, audioconvert_in, audioresample_in, caps_in, webrtc, adapter_out,
  //                  audioconvert_out, audioresample_out, caps_out, out_level, nullptr);

  // gst_element_link_many(in_level, adapter, audioconvert_in, audioresample_in, caps_in, webrtc, audioconvert_out,
  //                       audioresample_out, caps_out, adapter_out, out_level, nullptr);

  // auto* pad_sink = gst_element_get_static_pad(in_level, "sink");
  // auto* pad_src = gst_element_get_static_pad(out_level, "src");

  // gst_element_add_pad(bin, gst_ghost_pad_new("sink", pad_sink));
  // gst_element_add_pad(bin, gst_ghost_pad_new("src", pad_src));

  // gst_object_unref(GST_OBJECT(pad_sink));
  // gst_object_unref(GST_OBJECT(pad_src));

  // g_signal_connect(adapter, "notify::n-input-samples", G_CALLBACK(on_n_input_samples_changed), this);

  // g_settings_bind(settings, "post-messages", in_level, "post-messages", G_SETTINGS_BIND_DEFAULT);
  // g_settings_bind(settings, "post-messages", out_level, "post-messages", G_SETTINGS_BIND_DEFAULT);
}

void Webrtc::set_probe_input_node_id(const uint& id) const {
  // if (probe_src != nullptr) {
  //   auto path = std::to_string(id);

  //   g_object_set(probe_src, "path", path.c_str(), nullptr);

  //   util::debug(log_tag + "webrtc probe is using input device: " + path);
  // }
}

void Webrtc::bind_to_gsettings() {
  // g_settings_bind(settings, "high-pass-filter", webrtc, "high-pass-filter", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "echo-cancel", webrtc, "echo-cancel", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "echo-suppression-level", webrtc, "echo-suppression-level", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "noise-suppression", webrtc, "noise-suppression", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "noise-suppression-level", webrtc, "noise-suppression-level", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "gain-control", webrtc, "gain-control", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "extended-filter", webrtc, "extended-filter", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "delay-agnostic", webrtc, "delay-agnostic", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "target-level-dbfs", webrtc, "target-level-dbfs", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "compression-gain-db", webrtc, "compression-gain-db", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "limiter", webrtc, "limiter", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "voice-detection", webrtc, "voice-detection", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "gain-control-mode", webrtc, "gain-control-mode", G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "voice-detection-frame-size-ms", webrtc, "voice-detection-frame-size-ms",
  //                 G_SETTINGS_BIND_DEFAULT);

  // g_settings_bind(settings, "voice-detection-likelihood", webrtc, "voice-detection-likelihood",
  //                 G_SETTINGS_BIND_DEFAULT);
}

File: src/presets/webrtc.rs
use crate::presets::common::{update_key, update_string_key};
use serde::{Deserialize, Serialize};

use gtk::gio;
use gio::prelude::*;

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "kebab-case", default)]
pub struct WebRTC {
    state: bool,
    high_pass_filter: bool,
    echo_cancel: bool,
    echo_suppression_level: String,
    noise_suppression: bool,
    noise_suppression_level: String,
    gain_control: bool,
    extended_filter: bool,
    delay_agnostic: bool,
    target_level_dbfs: i32,
    compression_gain_db: i32,
    limiter: bool,
    gain_control_mode: String,
    voice_detection: bool,
    voice_detection_frame_size_ms: i32,
    voice_detection_likelihood: String,
}

impl Default for WebRTC {
    fn default() -> Self {
        let settings = gio::Settings::with_path(
            "com.github.wwmm.pulseeffects.webrtc",
            "/com/github/wwmm/pulseeffects/sourceoutputs/webrtc/",
        );

        WebRTC {
            state: settings.get_boolean("state"),
            high_pass_filter: settings.get_boolean("high-pass-filter"),
            echo_cancel: settings.get_boolean("echo-cancel"),
            echo_suppression_level: settings
                .get_string("echo-suppression-level")
                .unwrap()
                .to_string(),
            noise_suppression: settings.get_boolean("noise-suppression"),
            noise_suppression_level: settings
                .get_string("noise-suppression-level")
                .unwrap()
                .to_string(),
            gain_control: settings.get_boolean("gain-control"),
            extended_filter: settings.get_boolean("extended-filter"),
            delay_agnostic: settings.get_boolean("delay-agnostic"),
            target_level_dbfs: settings.get_int("target-level-dbfs"),
            compression_gain_db: settings.get_int("compression-gain-db"),
            limiter: settings.get_boolean("limiter"),
            gain_control_mode: settings
                .get_string("gain-control-mode")
                .unwrap()
                .to_string(),
            voice_detection: settings.get_boolean("voice-detection"),
            voice_detection_frame_size_ms: settings.get_int("voice-detection-frame-size-ms"),
            voice_detection_likelihood: settings
                .get_string("voice-detection-likelihood")
                .unwrap()
                .to_string(),
        }
    }
}

impl WebRTC {
    pub fn apply(&self) {
        let settings = gio::Settings::with_path(
            "com.github.wwmm.pulseeffects.webrtc",
            "/com/github/wwmm/pulseeffects/sourceoutputs/webrtc/",
        );

        update_key(&settings, "state", self.state);
        update_key(&settings, "high-pass-filter", self.high_pass_filter);
        update_key(&settings, "echo-cancel", self.echo_cancel);
        update_string_key(
            &settings,
            "echo-suppression-level",
            &self.echo_suppression_level,
        );
        update_key(&settings, "noise-suppression", self.noise_suppression);
        update_string_key(
            &settings,
            "noise-suppression-level",
            &self.noise_suppression_level,
        );
        update_key(&settings, "gain-control", self.gain_control);
        update_key(&settings, "extended-filter", self.extended_filter);
        update_key(&settings, "delay-agnostic", self.delay_agnostic);
        update_key(&settings, "target-level-dbfs", self.target_level_dbfs);
        update_key(&settings, "compression-gain-db", self.compression_gain_db);
        update_key(&settings, "limiter", self.limiter);
        update_string_key(&settings, "gain-control-mode", &self.gain_control_mode);
        update_key(&settings, "voice-detection", self.voice_detection);
        update_key(
            &settings,
            "voice-detection-frame-size-ms",
            self.voice_detection_frame_size_ms,
        );
        update_string_key(
            &settings,
            "voice-detection-likelihood",
            &self.voice_detection_likelihood,
        );
    }
}
